MystenLabs/sui
# [move-ide] Modified move-analyzer installation procedure to utilize versioning

## Description 

This goal of this PR is to allow friction-less updates of the Move analyzer binary, particularly for developers that utilize Sui CLI, which should be the majority. We still want friction-less Move analyzer binary installation for developers that want to only use the IDE and do not have a local CLI installation.

Move analyzer can now be installed in three different way:
1. it's bundled with the VSCode extension and if no local Move analyzer installation exists (or local installations have a lower version than the bundled one), it gets installed as part of VSCode Move extension activation process
2. it's bundled with Sui's CLI binary (if this binary has the highest version of all those installed, it will be used)
3. it can be built locally (or downloaded from GitHub) and placed in the default location (`~/.sui/bin`) or in a location specified by the user via VSCode extension's option

In cases 1 and 3 what gets ultimately used by the VSCode extension is the "standalone" Move analyzer binary (`move-analyzer` binary containing just the LSP server implementation). In case 2 what gets ultimately used is the Sui CLI which contains Move analyzer binary (and which is invoked via a special `analyzer` Sui CLI option).

Once this PR lands, the following will be enabled:
- if the first and only thing a developer does is to install VSCode extension from the marketplace, it's ready to use with all the functionality available
- if the developer installed Sui CLI, with each update of the CLI binary a new version of Move analyzer will be available even if the extension in the marketplace is not updated (which helps keeping parity between compiler version in Move analyzer binary and compiler version in the CLI binary, to avoid discrepancy with respect to features, error messages, etc.)

## Test plan 

Tested manually the following scenarios:

1. Check cases when no higher installed version exists:
- install bundled extension when no local installation exists
- use locally installed standalone version (default path) if bundled has the same version (no Sui version installed)
- use locally installed standalone version (user-specified path) if bundled has the same version (no Sui version installed)
- use locally installed Sui version (on system path) if bundled has the same version (no standalone version installed)
- use locally installed standalone version (user-specified path) if bundled has the same version and installed Sui one has the same version
- use locally installed Sui version (on user-specified path) if bundled has the same version (no standalone version installed)
- report error if no binary at user-specified path (no other local installs)
- report error if no binary at user-specified path (Sui version installed on system path)
2. Check cases when a higher installed version exists:
- use locally installed standalone version (default path) if bundled has lower version (no Sui version installed)
- use locally installed Sui version (on system path) if bundled has lower version (no standalone version installed)
3. Check cases when a lower installed version exists:
- install bundled version if installed standalone version is lower (no Sui version installed)
- install bundled version if installed Sui version is lower (no standalone version installed)
- install bundled version if both installed Sui version and standalone version are lower
